Data/models/CRUDProtocols.swift

```diff
@@ -0,0 +1,111 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import CoreData
+import Shared
+import XCGLogger
+
+private let log = Logger.browserLogger
+
+// TODO: Creatable, Updateable. Those are not needed at the moment.
+typealias CRUD = Readable & Deletable
+
+public protocol Deletable where Self: NSManagedObject {
+    func delete()
+    static func deleteAll(predicate: NSPredicate?)
+}
+
+public protocol Readable where Self: NSManagedObject {
+    static func count(predicate: NSPredicate?) -> Int?
+    static func first(where predicate: NSPredicate?, context: NSManagedObjectContext) -> Self?
+    static func all(where predicate: NSPredicate?, sortDescriptors: [NSSortDescriptor]?, fetchLimit: Int, 
+                       context: NSManagedObjectContext) -> [Self]?
+}
+
+// MARK: - Implementations
+public extension Deletable where Self: NSManagedObject {
+    func delete() {
+        let context = self.managedObjectContext ?? DataController.backgroundContext
+        
+        context.perform {
+            context.delete(self)
+            if !context.hasChanges {
+                log.debug("Wanting to remove object \(self) but context registered no changes.")
+                return
+            }
+            
+            do {
```

Comment:
May as well utilize the DataController's save mechanism. Maybe the context is worth the benefit of this location?